consensus/src/lib.rs

```diff
@@ -1,27 +1,1095 @@
+        use rand::prelude::*;
+
+        // For each output we create a new input and output that will placed into a new block.
+        // If value of original output is less than 1 then output will disappear in a new block.
+        // Otherwise, value will be decreasing for 1.
+        let (inputs, outputs): (Vec<TxInput>, Vec<TxOutput>) = prev_block
+            .transactions()
+            .iter()
+            .flat_map(|tx| {
+                let tx_id = tx.get_id();
+                tx.get_outputs()
+                    .iter()
+                    .enumerate()
+                    .filter_map(move |(index, output)| {
+                        if output.get_value() > Amount::from(1) {
+                            // Random address receiver
+                            let mut rng = rand::thread_rng();
+                            let mut witness: Vec<u8> = (1..100).collect();
+                            witness.shuffle(&mut rng);
+                            let mut address: Vec<u8> = (1..22).collect();
+                            address.shuffle(&mut rng);
+                            let receiver =
+                                Address::new(config, address).expect("Failed to create address");
+                            Some((
+                                TxInput::new(
+                                    OutPointSourceId::Transaction(tx_id.clone()),
+                                    index as u32,
+                                    witness,
+                                ),
+                                TxOutput::new(
+                                    (output.get_value() - Amount::from(1)).unwrap(),
+                                    Destination::Address(receiver),
+                                ),
+                            ))
+                        } else {
+                            None
+                        }
+                    })
+                    .collect::<Vec<(TxInput, TxOutput)>>()
+            })
+            .unzip();
+
+        Block::new(
+            vec![Transaction::new(0, inputs, outputs, 0).expect("Failed to create transaction")],
+            if orphan {
+                Some(Id::new(&H256::random()))
+            } else {
+                Some(Id::new(&prev_block.get_id().get()))
+            },
+            time::get() as u32,
+            ConsensusData::None,
+        )
+        .expect("Error creating block")
+    }
+
+    #[test]
+    #[allow(clippy::eq_op)]
+    fn test_process_genesis_block_wrong_block_source() {
+        // Genesis can't be from Peer, test it
+        let config = create_mainnet();
+        let storage = Store::new_empty().unwrap();
+        let mut consensus = Consensus::new(config.clone(), storage);
+
+        // process the genesis block
+        let block_source = BlockSource::Peer(0);
+        let result = consensus.process_block(config.genesis_block().clone(), block_source);
+        assert_eq!(result, Err(BlockError::InvalidBlockSource));
+    }
```

Comment:
A bunch of these tests fail when run under loom. Loom has to be set up first in order to work. There are two ways to resolve it:

1. Add support for loom for the test by wrapping its code in this construct:
   ```rust
   common::concurrency::model(|| {
       /* test code here */
   });
2. Disable given test or the whole `tests` module if run under `loom`:
   ```rust
   #[cfg(not(loom))]
   ```

Generally, I would favour running the tests with loom enabled unless it is not possible or there is too much concurrency for loom to finish in a reasonable amount of time.

Reply:
Thank you for explanation, Lukas! I've seen that these tests fails related to Loom, but I haven't seen this crate before.